pkg/julia/manifest/parse.go

```diff
@@ -0,0 +1,218 @@
+	var dependOn []string
+
+	for depName, depUUID := range deps {
+		dep, err := lookupDep(depName, depUUID, allDeps)
+		if err != nil {
+			return nil, err
+		}
+		version := depVersion(dep, juliaVersion)
+		dependOn = append(dependOn, utils.PackageID(dep.UUID, version))
+	}
+
+	if len(dependOn) > 0 {
+		sort.Strings(dependOn)
+		return &types.Dependency{
+			ID:        pkgId,
+			DependsOn: dependOn,
+		}, nil
+	} else {
+		return nil, nil
+	}
+}
+
+// Returns the matching dependency in `allDeps` given the dep with the `name` and `uuid`.
+// The `uuid` may be `nil` if the given dep is the only one with its name in the manifest.
+// Otherwise, if there are multiple deps with the same name, the `uuid` must be specified.
+func lookupDep(name string, uuid *string, allDeps map[string][]DecodedDependency) (*DecodedDependency, error) {
+	if uuid == nil {
+		// No UUID was set in the manifest, which means there is only one dep with this name
+		return &allDeps[name][0], nil
+	} else {
+		for _, candidateDep := range allDeps[name] {
+			if candidateDep.UUID == *uuid {
+				return &candidateDep, nil
+			}
+		}
+		return nil, xerrors.Errorf("failed to find dep with name %s and UUID %s", name, *uuid)
+	}
+}
+
+// Returns the effective version of the `dep`.
+// stdlib packages do not have a version in the manifest because they are packaged with julia itself
+func depVersion(dep *DecodedDependency, juliaVersion string) string {
+	if len(dep.Version) == 0 {
+		return juliaVersion
+	}
+	return dep.Version
+}
+
+// Decodes a primitive manifest using the metadata from parse time.
+func decodeManifest(man *PrimitiveManifest, metadata *toml.MetaData) (*DecodedManifest, error) {
+	// Copy over already decoded fields
+	decodedManifest := DecodedManifest{
+		JuliaVersion:   man.JuliaVersion,
+		ManifestFormat: man.ManifestFormat,
+		Dependencies:   make(map[string][]DecodedDependency),
+	}
+
+	// Decode each dependency into the new manifest
+	for depName, primDeps := range man.Dependencies {
+		decodedDeps := []DecodedDependency{}
+		for _, primDep := range primDeps {
+			decodedDep, err := decodeDependency(&primDep, metadata)
+			if err != nil {
+				return nil, err
+			}
+			decodedDeps = append(decodedDeps, *decodedDep)
+		}
+		decodedManifest.Dependencies[depName] = decodedDeps
+	}
+
+	return &decodedManifest, nil
+}
+
+// Decodes a primitive dependency using the metadata from parse time.
+func decodeDependency(dep *PrimitiveDependency, metadata *toml.MetaData) (*DecodedDependency, error) {
+	// Try to decode as []string first where the manifest looks like deps = ["A", "B"]
+	var possibleDeps []string
+	err := metadata.PrimitiveDecode(dep.Dependencies, &possibleDeps)
+	if err == nil {
+		finalDeps := make(map[string]*string)
+		for _, depName := range possibleDeps {
+			finalDeps[depName] = nil
+		}
+		return &DecodedDependency{
+			Dependencies: finalDeps,
+			UUID:         dep.UUID,
+			Version:      dep.Version,
+		}, nil
+	}
```

Comment:
I am not sure that we need both `PrimitiveDependency` and `DecodedDependency`.
What if we will add :
```
DependsOs    []string       `toml:"-"`       // list of dependent UUID's.
``` 
 field here - https://github.com/aquasecurity/go-dep-parser/blob/275db58029ca208af1928c0ab925929e37804444/pkg/julia/manifest/parse.go#L27-L31
and handle uuids for `["A", "B"]` case in this function.
Something like that:
```suggestion
// Decodes a primitive dependency using the metadata from parse time.
func decodeDependency(man *PrimitiveManifest, dep *PrimitiveDependency, metadata *toml.MetaData) (*DecodedDependency, error) {
	// Try to decode as []string first where the manifest looks like deps = ["A", "B"]
	var possibleDeps []string
	err := metadata.PrimitiveDecode(dep.Dependencies, &possibleDeps)
	if err == nil {
		var possibleUuids []string
		for _, depName := range possibleDeps {
			primDep := man.Dependencies[depName]
			if len(primDep) > 1 {
				// it is case for [deps.A.deps]
				// return error
			}
			possibleUuids = append(possibleUuids, primDep[0].UUID)
		}
		dep.DependsOs = possibleUuids
		return dep, nil
	}
```
`[deps.A.deps]` case already uses UUID

Reply:
That definition of `PrimitiveDependency` can't get parsed: `toml: line 10 (last key "deps.A.deps"): incompatible types: TOML value has type map[string]interface {}; destination has type slice`. That error is why I delayed the parsing until later in `decodeDependency`.

Reply:
Perhaps i confused you.
I meant that we stay your solution for `PrimitiveDependency`, but add 1 more field - https://gist.github.com/DmitriyLewen/22431f167fdb110b5b17a90b6071051b#file-parse-go-L28-L33

Also we need to use `PrimitiveDecode` twice ( as you did that):
https://gist.github.com/DmitriyLewen/22431f167fdb110b5b17a90b6071051b#file-parse-go-L140-L172

My code in gist needs refactoring, but i hope you will understand my logic.
PS. sorry, i didn't pull your changes for ID.